### ISSUE TITLE
"Sampler and server are owned by the same process" >>> hangs

### ISSUE THREAD

**Issue by kai-luni** [PatternsandPredictions/DRLearner_beta#2]
Hi guys, thanks for this great application, I am a huge fan.

Unfortunately it gets stuck when I use

`python ./examples/run_atari.py --level PongNoFrameskip-v4 --num_episodes 1000 --exp_path experiments/test_pong/`

The output is 

`I0218 12:13:59.176958 139760901358784 savers.py:166] Attempting to restore checkpoint: experiments/test_pong/checkpoints/learner/ckpt-28
I0218 12:14:01.921640 139760901358784 csv.py:76] Logging to experiments/test_pong/ffea80e2-af4a-11ed-a68c-00155dab5046/logs/learner/logs.csv
I0218 12:14:04.757199 139760901358784 savers.py:156] Saving checkpoint: experiments/test_pong/checkpoints/learner
[Learner] Action Mean Time = 0.013 | Env Step Mean Time = 0.005 | Episode Length = 765 | Episode Return = -21.0 | Episodes = 1 | Observe Mean Time = 0.012 | Steps = 765 | Steps Per Second = 27.112
[Learner] Action Mean Time = 0.009 | Env Step Mean Time = 0.004 | Episode Length = 825 | Episode Return = -21.0 | Episodes = 2 | Observe Mean Time = 0.008 | Steps = 1590 | Steps Per Second = 45.701
[Learner] Action Mean Time = 0.007 | Env Step Mean Time = 0.003 | Episode Length = 765 | Episode Return = -21.0 | Episodes = 3 | Observe Mean Time = 0.007 | Steps = 2355 | Steps Per Second = 56.890
[Learner] Action Mean Time = 0.006 | Env Step Mean Time = 0.003 | Episode Length = 903 | Episode Return = -20.0 | Episodes = 4 | Observe Mean Time = 0.006 | Steps = 3258 | Steps Per Second = 65.528
[Learner] Action Mean Time = 0.007 | Env Step Mean Time = 0.003 | Episode Length = 811 | Episode Return = -21.0 | Episodes = 5 | Observe Mean Time = 0.007 | Steps = 4069 | Steps Per Second = 58.085
[Learner] Action Mean Time = 0.007 | Env Step Mean Time = 0.003 | Episode Length = 842 | Episode Return = -20.0 | Episodes = 6 | Observe Mean Time = 0.007 | Steps = 4911 | Steps Per Second = 58.781
[Learner] Action Mean Time = 0.007 | Env Step Mean Time = 0.003 | Episode Length = 765 | Episode Return = -21.0 | Episodes = 7 | Observe Mean Time = 0.007 | Steps = 5676 | Steps Per Second = 55.588
[reverb/cc/client.cc:165] Sampler and server are owned by the same process (604) so Table priority_table is accessed directly without gRPC.
[reverb/cc/client.cc:165] Sampler and server are owned by the same process (604) so Table priority_table is accessed directly without gRPC.
[reverb/cc/client.cc:165] Sampler and server are owned by the same process (604) so Table priority_table is accessed directly without gRPC.
[reverb/cc/client.cc:165] Sampler and server are owned by the same process (604) so Table priority_table is accessed directly without gRPC.
[reverb/cc/client.cc:165] Sampler and server are owned by the same process (604) so Table priority_table is accessed directly without gRPC.`

And then nothing happens anymore. I suspect that reverb is blocking.

I tried to change the rate limiter in builder.py but I didnt really know what I was doing.

I followed the instructions closely with python 3.7, conda env and everything. I am using Ubuntu 22 on wsl. Do you have any recommendation what I can try?

**Comment by ghharrison** [comment]
I have encountered the same problem, also using Python 3.7 and Miniconda for the env, using Ubuntu on WSL. Curiously, the hanging happens at the same place, after 7 episodes of training. However, after an initial hang LunarLander seems able to work.

**Comment by hlsafin** [comment]
Yes, I get the same issue now :( Lunar Lander works but not Pong. Any update on this?

**Comment by DamirHadiiev** [comment]
Hi! This issue appears only in the mode of synchronous run because of deadlock related to reverb (basically examples, where there is no prefix **distrun** in examples folder). Best efficiency of this library is guarantied at distrun, though (the one you will run to play large games, like Montezuma).

You can actually make things work by setting with samples_per_rate != 0 in the corresponding config, and it will run.

**Comment by kai-luni** [comment]
Thanks that you took the time, I will try it.

**Comment by DamirHadiiev** [comment]
In the latest version, there is a warning in sync mode of run_atari.py to avoid confusion 😄 
 https://github.com/PatternsandPredictions/DRLearner_beta/blob/e5f1eddb72f2723c260917f302c04d7a4a2881a3/examples/run_atari.py#L58
